src/core/sphinx/peel.rs

```diff
@@ -0,0 +1,190 @@
+	DeliverCover { cover_id: Option<CoverId> },
+}
+
+#[derive(Debug, thiserror::Error, PartialEq, Eq)]
+pub enum PeelErr {
+	#[error(
+		"Bad key-exchange public key in header (Diffie-Hellman exchange produced identity point)"
+	)]
+	KxPublic,
+	#[error("Bad MAC in header")]
+	Mac,
+	#[error("Bad payload tag")]
+	PayloadTag,
+}
+
+fn check_payload_tag(tag: &PayloadTag) -> Result<(), PeelErr> {
+	let tag_ok: bool = tag.ct_eq(&PAYLOAD_TAG).into();
+	if tag_ok {
+		Ok(())
+	} else {
+		Err(PeelErr::PayloadTag)
+	}
+}
+
+/// Attempt to peel a layer off `packet` using `kx_shared_secret`. `kx_shared_secret` should be
+/// derived from `kx_public(packet)` and this node's secret key.
+pub fn peel(
+	out: &mut Packet,
+	packet: &Packet,
+	kx_shared_secret: &KxSharedSecret,
+) -> Result<Action, PeelErr> {
+	// This can only happen in the case of a malicious message sender. I don't believe it would be
+	// a problem to skip this check, however it's cheap and the Sphinx paper says to do it.
+	if kx_shared_secret_is_identity(kx_shared_secret) {
+		return Err(PeelErr::KxPublic)
+	}
+
+	// (kx_public, mac, encrypted_header, payload) correspond to (alpha, gamma, beta, delta) in the
+	// Sphinx paper
+	let (kx_public, mac, encrypted_header, payload) =
+		array_refs![packet, KX_PUBLIC_SIZE, MAC_SIZE, ENCRYPTED_HEADER_SIZE, PAYLOAD_SIZE];
+
+	let sds = SmallDerivedSecrets::new(kx_shared_secret);
+
+	// Verify the MAC
+	if !mac_ok(mac, encrypted_header, sds.mac_key()) {
+		return Err(PeelErr::Mac)
+	}
+
+	// Decrypt the header and generate padding for length invariance. Try to get the decrypted
+	// header in the right place in the output. The most likely case is that we will be forwarding
+	// the packet to another node, so assume this. We could save some work in the deliver case by
+	// decrypting just the first few bytes to start with to see if we need to decrypt the rest.
+	// This would complicate things and as the forward case is much more common it might ultimately
+	// not make things any faster, so don't bother for now.
+	let decrypted_header = array_mut_ref![
+		out,
+		KX_PUBLIC_SIZE - RAW_ACTION_SIZE,
+		ENCRYPTED_HEADER_SIZE + MAX_HEADER_PAD_SIZE
+	];
+	*array_mut_ref![decrypted_header, 0, ENCRYPTED_HEADER_SIZE] = *encrypted_header;
+	*array_mut_ref![decrypted_header, ENCRYPTED_HEADER_SIZE, MAX_HEADER_PAD_SIZE] =
+		[0; MAX_HEADER_PAD_SIZE]; // Padding is generated by encrypting zeroes
+	apply_header_encryption_keystream(decrypted_header, sds.header_encryption_key());
+
+	let raw_action = RawAction::from_le_bytes(*array_ref![decrypted_header, 0, RAW_ACTION_SIZE]);
+	Ok(match raw_action {
+		RAW_ACTION_DELIVER_REQUEST => {
+			// Peel off the final layer of payload encryption
+			let out = array_mut_ref![out, 0, PAYLOAD_SIZE];
+			*out = *payload;
+			decrypt_payload(out, &derive_payload_encryption_key(kx_shared_secret));
+
+			check_payload_tag(array_ref![out, PAYLOAD_DATA_SIZE, PAYLOAD_TAG_SIZE])?;
```

Comment:
I moved the tag after the data to keep the `peel` and `decrypt_reply_payload` interfaces simple while avoiding extra data movement. If the tag was at the start then either the interfaces would be a bit weird (because you would need to ignore the first `PAYLOAD_TAG_SIZE` bytes of the output), or we would have to shuffle the data down after decryption.

Reply:
The question is what is the purpose of this tag. I don't know for sure, but my opinion was to avoid deciphering the whole payload if incorrect (actually was not even implemented this way previously).
Probably not an issue, maybe worth mentioning in doc (if it is not already).

Reply:
then it is just better at the end for chained cypher.